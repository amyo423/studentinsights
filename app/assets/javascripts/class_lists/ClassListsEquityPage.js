import React from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import School from '../components/School';
import GenericLoader from '../components/GenericLoader';
import SectionHeading from '../components/SectionHeading';
import ExperimentalBanner from '../components/ExperimentalBanner';
import tableStyles from '../components/tableStyles';
import {gradeText} from '../helpers/gradeText';
import {rankedByGradeLevel} from '../helpers/SortHelpers';
import {fetchExperimentalWorkspacesWithEquity} from './api';


// Experimental view for looking at equity in class lists
export default class ClassListsEquityPage extends React.Component {
  render() {
    const {currentEducatorId} = this.props;
    return (
      <div className="ClassListsEquityPage">
        <GenericLoader
          style={styles.root}
          promiseFn={fetchExperimentalWorkspacesWithEquity}
          render={json => (
            <ClassListsEquityPageView
              currentEducatorId={currentEducatorId}
              dimensionKeys={json.dimension_keys}
              classListsWithDimensions={json.class_lists_with_dimensions} />
          )} />
      </div>
    );
  }
}
ClassListsEquityPage.propTypes = {
  currentEducatorId: PropTypes.number.isRequired
};

// View component
export class ClassListsEquityPageView extends React.Component {
  render() {
    return (
      <div>
        <ExperimentalBanner />
        <SectionHeading>Class List: Experimental Equity View</SectionHeading>
        {this.renderTable()}
      </div>
    );
  }

  renderTable() {
    const {dimensionKeys, classListsWithDimensions, currentEducatorId} = this.props;

    if (classListsWithDimensions.length === 0) return <div>No class lists</div>;

    const sortedClassLists = _.sortByOrder(classListsWithDimensions, classList => {
      return [
        classList.school.name,
        rankedByGradeLevel(classList.grade_level_next_year),
        classList.submitted
      ];
    });

    const headerCell = {...tableStyles.headerCell, fontSize: 10};
    const cell = {...tableStyles.cell, textAlign: 'center'};
    return (
      <div>
        <table style={tableStyles.table}>
          <thead>
            <tr>
              <th style={headerCell}>School</th>
              <th style={headerCell}>Grade next year</th>
              {dimensionKeys.map(dimensionKey => (
                <th key={dimensionKey} style={headerCell}>{dimensionKey}</th>
              ))}
              <th style={headerCell} />
            </tr>
          </thead>
          <tbody>{sortedClassLists.map(classList => {
            const educatorStyle = (classList.created_by_teacher_educator.id === currentEducatorId)
              ? { fontWeight: 'bold' }
              : {};
            return (
              <tr key={classList.workspace_id}>
                <td style={cell}><School {...classList.school} /></td>
                <td style={cell}>
                  {gradeText(classList.grade_level_next_year)}
                </td>
                {dimensionKeys.map(dimensionKey => {
                  const dimension = _.find(classList.dimensions, { dimension_key: dimensionKey });
                  const value = dimension.index.herfindahl;
                  return <td key={dimensionKey} style={cell}>{this.renderValue(value)}</td>;
                })}
                <td style={cell}>
                  <a style={{padding: 10}} href={`/classlists/${classList.workspace_id}`}>open</a>
                </td>
              </tr>
            );
          })}</tbody>
        </table>
      </div>
    );
  }

  renderValue(value) {
    const rounded = Math.round(value * 100);
    if (value < .25) {
      return <span style={{padding: 10, color: 'white', backgroundColor: 'green', fontWeight: 'bold'}}>{rounded}</span>;
    } else if (value < .60) {
      return <span style={{padding: 10, color: 'black', backgroundColor: 'white', fontWeight: 'normal'}}>{rounded}</span>;
    } else {
      return <span style={{padding: 10, color: 'white', backgroundColor: 'red', fontWeight: 'bold'}}>{rounded}</span>;
    }
  }
}
ClassListsEquityPageView.propTypes = {
  currentEducatorId: PropTypes.number.isRequired,
  dimensionKeys: PropTypes.arrayOf(PropTypes.string).isRequired,
  classListsWithDimensions: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number.isRequired,
    workspace_id: PropTypes.string.isRequired,
    grade_level_next_year: PropTypes.string.isRequired,
    created_at: PropTypes.string.isRequired,
    updated_at: PropTypes.string.isRequired,
    submitted: PropTypes.bool.isRequired,
    created_by_teacher_educator: PropTypes.object.isRequired,
    school: PropTypes.object.isRequired,
    dimensions: PropTypes.arrayOf(PropTypes.shape({
      dimension_key: PropTypes.string.isRequired,
      index: PropTypes.object.isRequired
    })).isRequired
  })).isRequired
};


const styles = {
  root: {
    padding: 10
  },
  newButton: {
    display: 'block',
    marginTop: 10
  },
  overview: {
    margin: 10
  },
  p: {
    marginBottom: 10
  }
};
