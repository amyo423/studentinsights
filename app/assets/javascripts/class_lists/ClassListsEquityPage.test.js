import React from 'react';
import ReactDOM from 'react-dom';
import fetchMock from 'fetch-mock/es5/client';
import renderer from 'react-test-renderer';
import ClassListsEquityPage, {ClassListsEquityPageView} from './ClassListsEquityPage';
import experimental_workspaces_with_equity_json from './fixtures/experimental_workspaces_with_equity_json';

beforeEach(() => {
  fetchMock.reset();
  fetchMock.restore();
  fetchMock.get('express:/api/class_lists/experimental_workspaces_with_equity_json', experimental_workspaces_with_equity_json);
});

export function testProps(props = {}) {
  return {
    currentEducatorId: 999999,
    ...props
  };
}

it('renders without crashing', () => {
  const el = document.createElement('div');
  const props = testProps();
  ReactDOM.render(<ClassListsEquityPage {...props} />, el);
});


it('snapshots view', () => {
  const json = experimental_workspaces_with_equity_json;
  const props = testProps({
    dimensionKeys: json.dimension_keys,
    classListsWithDimensions: json.class_lists_with_dimensions
  });
  const tree = renderer
    .create(<ClassListsEquityPageView {...props} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});