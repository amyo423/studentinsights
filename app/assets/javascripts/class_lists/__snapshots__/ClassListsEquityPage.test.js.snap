// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots view 1`] = `
<div>
  <div
    style={
      Object {
        "background": "red",
        "color": "white",
        "fontWeight": "bold",
        "padding": 20,
      }
    }
  >
    <span>
      This is an experimental prototype page!
    </span>
    <span
      style={
        Object {
          "marginLeft": 10,
        }
      }
    >
      Please try it out and share your feedback.
    </span>
  </div>
  <div
    className="SectionHeading"
    style={
      Object {
        "borderBottom": "1px solid #333",
        "padding": 10,
      }
    }
  >
    <h4
      style={
        Object {
          "color": "black",
          "display": "inline",
        }
      }
    >
      Class List: Experimental Equity View
    </h4>
  </div>
  <div>
    <table
      style={
        Object {
          "border": "1px solid #eee",
          "borderCollapse": "collapse",
          "margin": 10,
        }
      }
    >
      <thead>
        <tr>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            School
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            Grade next year
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            iep_or_504
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            limited_or_flep
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            male
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            reduced_lunch
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            high_discipline
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            whiteness
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          >
            color
          </th>
          <th
            style={
              Object {
                "backgroundColor": "#ccc",
                "border": "1px solid #eee",
                "fontSize": 10,
                "fontWeight": "bold",
                "padding": 10,
                "textAlign": "left",
                "verticalAlign": "top",
              }
            }
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <a
              className="School"
              href="/schools/2"
              style={Object {}}
            >
              Arthur D Healey
            </a>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            6th grade
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "red",
                  "color": "white",
                  "fontWeight": "bold",
                  "padding": 10,
                }
              }
            >
              100
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "white",
                  "color": "black",
                  "fontWeight": "normal",
                  "padding": 10,
                }
              }
            >
              56
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "green",
                  "color": "white",
                  "fontWeight": "bold",
                  "padding": 10,
                }
              }
            >
              0
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "red",
                  "color": "white",
                  "fontWeight": "bold",
                  "padding": 10,
                }
              }
            >
              100
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "green",
                  "color": "white",
                  "fontWeight": "bold",
                  "padding": 10,
                }
              }
            >
              0
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "red",
                  "color": "white",
                  "fontWeight": "bold",
                  "padding": 10,
                }
              }
            >
              100
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <span
              style={
                Object {
                  "backgroundColor": "white",
                  "color": "black",
                  "fontWeight": "normal",
                  "padding": 10,
                }
              }
            >
              50
            </span>
          </td>
          <td
            style={
              Object {
                "border": "1px solid #eee",
                "padding": 10,
                "textAlign": "center",
                "verticalAlign": "top",
              }
            }
          >
            <a
              href="/classlists/af0cabfe-35c5-4c14-b6b8-2b725bf22e2b"
              style={
                Object {
                  "padding": 10,
                }
              }
            >
              open
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
